fix: auto-replace for pinDigest without currentDigest or shortDigest

setchy opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

This PR handles two new scenarios when using auto-replacement rules

**Scenario 1**
- `pinDigests` are enabled
-  dependency targeted for replacement does not have a `currentDigest`
  
 **Scenario 2**
- `digestShort` SHAs are being used
 
In both scnearios, the [verification logic](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts#L82) deems these as invalid and therefore no branches are created/updated.

I have validated these changes locally running about our Bitbucket Cloud test repository

<!-- Describe what behavior is changed by this PR. -->

## Context

Resolves https://github.com/renovatebot/renovate/issues/20128

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice commented on lib/workers/repository/update/branch/auto-replace.ts:
```diff
@@ -198,11 +199,15 @@ export async function doAutoReplace(
+            newDigest
+          );
+        } else if (pinDigests) {
+          newString = `${newString}@${newDigest}`;
```
this only works for docker images and will break all other managers with different dep types.

setchy replied:
@viceice - good feedback.  

do you have a suggestion on how to best proceed... would an additional guard make sense to check if it's a docker update?  

alternatively, perhaps `confirmIfDepUpdated` function could be updated to better handle this edge-case

setchy replied:
I've pivoted to a different solution to handle this edge-case.  explanation added as a comment below

setchy commented:
I've pivoted to a different solution thanks to @viceice initial feedback.

In this new method, replacements without a `currentDigest` will not attempt to add a `newDigest` in the same branch/PR, while replacements with a `currentDigest` will ensure its updated tothe `newDigest` value

viceice reviewed (commented):
looks good so far, can you make a more complex non-docker regex manager sample?

setchy commented:
I've added two `regex manager` tests.



_An observation I had while doing so....  I had some "fun" due to the test data I originally selected and the way in which the existing implementation of [version replacement ](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts#L189) and [digest replacement](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts#L201) is handled._

_For example, I began with a test case to replace `image:1@sha256:1a2b` with `new-image:3.14@sha256:9p8o`.  What ended up happening after the version replacement step is a string of `image:3.14@sha256:3.14a2b" which then skips the digest replacement step, which ultimately gets caught in the final `confirmIfDepUpdated` check.  I think fixing this interesting edge-case (if at all) should be in a separate issue/PR_

viceice commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -909,7 +909,7 @@ describe('workers/repository/update/branch/auto-replace', () => {
         module "vpc" {
           source  = "terraform-aws-modules/vpc/aws"
-         version = "3.14.2"
+          version = "3.14.2"
```
?

setchy replied:
The indentation was off

viceice reviewed (commented):
please add a GitHub actions test. there we've digest pinning too.

setchy commented:
> please add a GitHub actions test. there we've digest pinning too.

github actions / github-tags tests added

viceice commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -1038,5 +1038,191 @@ describe('workers/repository/update/branch/auto-replace', () => {
+      upgrade.depIndex = 0;
+      upgrade.pinDigests = true;
+      upgrade.updateType = 'replacement';
+      upgrade.replaceString = 'actions/checkout@v1.0.0';
```
i think you need the auto-replace template which is normally set by the github actions manager 🤔

setchy replied:
Added to the two github-actions replacement tests, although it won't be used due to [this conditional check](https://github.com/renovatebot/renovate/blob/51107e31608f81111774cf6aee77a1d2074fe7a3/lib/workers/repository/update/branch/auto-replace.ts#L185)

Furthermore, added two other github-actions tests to cover scenarios where we're not changing the action name

viceice commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -1038,5 +1038,284 @@ describe('workers/repository/update/branch/auto-replace', () => {
+
+    it('github-actions: update with newValue only', async () => {
+      const githubaction = codeBlock`
+        name: build
+
+        on: [push]
+
```
i think you can remove this, not relevant for test. so test is a little bit less verbose and big

setchy replied:
Done ✅

viceice commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -1038,5 +1038,284 @@ describe('workers/repository/update/branch/auto-replace', () => {
+          build:
+            runs-on: ubuntu-latest
+            steps:
+              - uses: actions/checkout@2485f4d55aae6c5b073114bc4c4b1907c0abae14166281beee7d93f76ebf41fc # tag=v1.0.0
```
why are the git sha so long? can we shorten them for better readability?

setchy replied:
Because the github-actions manager validates either [sha](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/github-actions/extract.ts#L17) or [shaShort](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/github-actions/extract.ts#L18)

setchy replied:
Done.  In doing so, I uncovered that the auto-replace didn't currently support digestShort updates.  This is now resolved too

setchy commented:
As part of updating the github-actions tests to use shortDigest formats (for readability), it was discovered that this scenario was also not handled by the auto-replace logic.  As such, I've updated the tests and added the necessary replacement step for shortDigests.  PR title has been renamed to represent the work

viceice commented on lib/workers/repository/update/branch/auto-replace.ts:
```diff
@@ -204,6 +213,12 @@ export async function doAutoReplace(
           newDigest
         );
       }
+      if (currentDigestShort && newDigest) {
```
```suggestion
      else if (currentDigestShort && newDigest) {
```

setchy replied:
Good suggestion 👍

viceice commented:
needs some more 👀👀👀

secustor commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -1038,5 +1038,250 @@ describe('workers/repository/update/branch/auto-replace', () => {
+      const res = await doAutoReplace(upgrade, dockerfile, reuseExistingBranch);
+      expect(res).toBe(
+        codeBlock`
+          FROM alpine:3.16
```
Shouldn't `pinDigests: true` lead to a digest even if there not one currently applied?
```suggestion
          FROM alpine:3.16@sha256:p0o9i8u7z6t5r4e3w2q1
```

setchy replied:
@secustor - thanks for the PR review.

I called out this behavior in this [comment above](https://github.com/renovatebot/renovate/pull/20317#issuecomment-1425112361) after feedback from @viceice [here](https://github.com/renovatebot/renovate/pull/20317#discussion_r1101824798)

secustor replied:
The missing `autoReplaceStringTemplate` which is added by the Dockerfile manager is the reason why . 
https://github.com/renovatebot/renovate/blob/90694bf398fea2f20c4892db48f61a6f420f8610/lib/modules/manager/dockerfile/extract.ts#L33

setchy replied:
Thanks @secustor - will update the test cases later today

setchy replied:
Ok, I've double confirmed what I mentioned above , adding `upgrade.autoReplaceStringTemplate=
        '{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}';` has no bearing with replacements because of how the [logic is inside doAutoReplace](https://github.com/renovatebot/renovate/blob/8af82990ee38f4e35071af01c8d124f401fa2bc1/lib/workers/repository/update/branch/auto-replace.ts#L196)

secustor commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -1038,5 +1038,250 @@ describe('workers/repository/update/branch/auto-replace', () => {
+        'image:\\s*?\\\'?\\"?(?<depName>[^:\\\'\\"]+):(?<currentValue>[^@\\\'\\"]+)@?(?<currentDigest>[^\\s\\\'\\"]+)?\\"?\\\'?\\s*',
+      ];
+      const res = await doAutoReplace(upgrade, yml, reuseExistingBranch);
+      expect(res).toBe('image: "some.other.url.com/some-new-repo:3.16"');
```
Same here

secustor replied:
Please also add an example with a set `autoReplaceStringTemplate` where the digest is added.

setchy replied:
As above

secustor commented on lib/workers/repository/update/branch/auto-replace.spec.ts:
```diff
@@ -1038,5 +1038,250 @@ describe('workers/repository/update/branch/auto-replace', () => {
+            build:
+              runs-on: ubuntu-latest
+              steps:
+                - uses: some-other-action/checkout@v2.0.0
```
Here too ☝️

JamieMagee reviewed (approved):
This looks great 🎉

chezsmithy commented:
This should resolve #19999 as well.
